File: engines/go/inserter_pgsql.go
package main

import (
	"context"
	"database/sql"
	"fmt"
	"os"
	"strings"

	"github.com/jackc/pgx/v5"
	_ "github.com/jackc/pgx/v5/stdlib"
)

type PgSQLInserter struct {
	db     *sql.DB
	params ConnParams
}

func NewPgSQLInserter(params ConnParams) (*PgSQLInserter, error) {
	dsn := fmt.Sprintf(
		"host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
		params.Host, params.Port, params.User, params.Password, params.Database,
	)
	db, err := sql.Open("pgx", dsn)
	if err != nil {
		return nil, fmt.Errorf("sql.Open: %w", err)
	}
	if err := db.Ping(); err != nil {
		return nil, fmt.Errorf("ping: %w", err)
	}
	return &PgSQLInserter{db: db, params: params}, nil
}

func (ins *PgSQLInserter) Close() {
	ins.db.Close()
}

func (ins *PgSQLInserter) CountRows(tableName string) (int, error) {
	var n int
	err := ins.db.QueryRow(fmt.Sprintf("SELECT COUNT(*) FROM %s", ins.quote(tableName))).Scan(&n)
	if err != nil {
		return 0, fmt.Errorf("count rows: %w", err)
	}
	return n, nil
}

func (ins *PgSQLInserter) quote(name string) string {
	clean := cleanStr(name)
	return `"` + strings.ReplaceAll(clean, `"`, `""`) + `"`
}

func (ins *PgSQLInserter) placeholder(i int) string {
	return fmt.Sprintf("$%d", i+1)
}

func (ins *PgSQLInserter) DefaultInsert(csvFile, tableName string) (int, error) {
	stream, err := openCSV(csvFile)
	if err != nil {
		return 0, err
	}
	defer stream.Close()

	cols := make([]string, len(stream.Headers))
	phs  := make([]string, len(stream.Headers))
	for i, h := range stream.Headers {
		cols[i] = ins.quote(h)
		phs[i]  = ins.placeholder(i)
	}

	query := fmt.Sprintf(
		"INSERT INTO %s (%s) VALUES (%s)",
		ins.quote(tableName),
		strings.Join(cols, ", "),
		strings.Join(phs, ", "),
	)

	tx, err := ins.db.Begin()
	if err != nil {
		return 0, fmt.Errorf("begin tx: %w", err)
	}
	stmt, err := tx.Prepare(query)
	if err != nil {
		tx.Rollback()
		return 0, fmt.Errorf("prepare stmt: %w", err)
	}
	defer stmt.Close()

	count := 0
	for {
		row, ok, err := stream.Next()
		if err != nil {
			tx.Rollback()
			return 0, err
		}
		if !ok {
			break
		}
		if _, err := stmt.Exec(rowToArgs(row)...); err != nil {
			tx.Rollback()
			return 0, fmt.Errorf("exec row: %w", err)
		}
		count++
	}

	return count, tx.Commit()
}

func (ins *PgSQLInserter) BulkInsert(csvFile, tableName string, batchSize int) (int, error) {
	stream, err := openCSV(csvFile)
	if err != nil {
		return 0, err
	}
	defer stream.Close()

	cols := make([]string, len(stream.Headers))
	for i, h := range stream.Headers {
		cols[i] = ins.quote(h)
	}
	colStr := strings.Join(cols, ", ")
	table  := ins.quote(tableName)
	ncols  := len(stream.Headers)
	total  := 0

	batch := make([][]string, 0, batchSize)

	flushBatch := func() error {
		if len(batch) == 0 {
			return nil
		}
		// PostgreSQL использует $N плейсхолдеры с глобальной нумерацией —
		// строим её на лету для каждой пачки.
		valueStrings := make([]string, len(batch))
		valueArgs    := make([]interface{}, 0, len(batch)*ncols)
		for j, row := range batch {
			phs := make([]string, ncols)
			for k := 0; k < ncols; k++ {
				phs[k] = ins.placeholder(j*ncols + k)
			}
			valueStrings[j] = "(" + strings.Join(phs, ", ") + ")"
			valueArgs = append(valueArgs, rowToArgs(row)...)
		}
		query := fmt.Sprintf(
			"INSERT INTO %s (%s) VALUES %s",
			table, colStr, strings.Join(valueStrings, ", "),
		)
		tx, err := ins.db.Begin()
		if err != nil {
			return fmt.Errorf("begin tx: %w", err)
		}
		if _, err := tx.Exec(query, valueArgs...); err != nil {
			tx.Rollback()
			return fmt.Errorf("exec batch: %w", err)
		}
		if err := tx.Commit(); err != nil {
			return fmt.Errorf("commit: %w", err)
		}
		total += len(batch)
		batch = batch[:0]
		return nil
	}

	for {
		row, ok, err := stream.Next()
		if err != nil {
			return total, err
		}
		if !ok {
			break
		}
		batch = append(batch, row)
		if len(batch) >= batchSize {
			if err := flushBatch(); err != nil {
				return total, err
			}
		}
	}
	if err := flushBatch(); err != nil {
		return total, err
	}

	return total, nil
}

func (ins *PgSQLInserter) FileInsert(csvFile, tableName string) (int, error) {
	// Файл не парсим в Go вообще — стримим напрямую через COPY FROM STDIN.
	// PostgreSQL сам понимает CSV (RFC 4180). RowsAffected возвращает
	// фактическое число вставленных строк.
	ctx := context.Background()
	dsn := fmt.Sprintf(
		"host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
		ins.params.Host, ins.params.Port,
		ins.params.User, ins.params.Password, ins.params.Database,
	)
	conn, err := pgx.Connect(ctx, dsn)
	if err != nil {
		return 0, fmt.Errorf("pgx connect: %w", err)
	}
	defer conn.Close(ctx)

	f, err := os.Open(csvFile)
	if err != nil {
		return 0, fmt.Errorf("open csv: %w", err)
	}
	defer f.Close()

	copySQL := fmt.Sprintf(
		"COPY %s FROM STDIN WITH (FORMAT csv, HEADER true)",
		ins.quote(tableName),
	)

	tag, err := conn.PgConn().CopyFrom(ctx, f, copySQL)
	if err != nil {
		return 0, fmt.Errorf("copy from stdin: %w", err)
	}
	return int(tag.RowsAffected()), nil
}

File: engines/rust/src/inserter_pgsql.rs
use crate::csv::{CsvStream, clean_identifier};
use crate::inserter::{ConnParams, Inserter};
use anyhow::{Result, anyhow};
use postgres::{Client, NoTls};
use std::fmt::Write;
use std::io::BufRead;

pub struct PgSQLInserter {
    client: Client,
}

impl PgSQLInserter {
    pub fn new(p: &ConnParams) -> Result<Self> {
        let dsn = format!(
            "host={} port={} user={} password={} dbname={} sslmode=disable",
            p.host, p.port, p.user, p.password, p.database
        );
        let client = Client::connect(&dsn, NoTls)?;
        Ok(Self { client })
    }

    fn quote(name: &str) -> String {
        let clean = clean_identifier(name);
        format!("\"{}\"", clean.replace('"', "\"\""))
    }

    fn build_cols(headers: &[String]) -> String {
        headers.iter()
            .map(|h| Self::quote(h))
            .collect::<Vec<_>>()
            .join(", ")
    }

    fn build_placeholders(count: usize) -> String {
        (1..=count)
            .map(|i| format!("${}", i))
            .collect::<Vec<_>>()
            .join(", ")
    }
}

impl Inserter for PgSQLInserter {

    fn default_insert(&mut self, csv_file: &str, table: &str) -> Result<usize> {
        let mut stream = CsvStream::open(csv_file)?;
        let ncols = stream.headers.len();
        let sql   = format!(
            "INSERT INTO {} ({}) VALUES ({})",
            Self::quote(table),
            Self::build_cols(&stream.headers),
            Self::build_placeholders(ncols),
        );

        let stmt = self.client.prepare(&sql)?;
        let mut tx = self.client.transaction()?;
        let mut count = 0;

        while let Some(row) = stream.next() {
            let row = row?;
            let params: Vec<&(dyn postgres::types::ToSql + Sync)> = row.iter()
                .map(|v| v as &(dyn postgres::types::ToSql + Sync))
                .collect();
            tx.execute(&stmt, &params)?;
            count += 1;
        }
        tx.commit()?;
        Ok(count)
    }

    fn bulk_insert(&mut self, csv_file: &str, table: &str,
                   batch_size: usize) -> Result<usize> {
        let mut stream = CsvStream::open(csv_file)?;
        let ncols  = stream.headers.len();
        let cols   = Self::build_cols(&stream.headers);
        let qtable = Self::quote(table);
        let mut total = 0;

        let mut batch: Vec<Vec<String>> = Vec::with_capacity(batch_size);
        let mut eof = false;

        while !eof {
            match stream.next() {
                Some(Ok(row)) => batch.push(row),
                Some(Err(e))  => return Err(e),
                None          => eof = true,
            }

            if (batch.len() >= batch_size || (eof && !batch.is_empty())) && !batch.is_empty() {
                // Собираем (..., $i, $i+1, ...), ($j, ...) — нумерация
                // плейсхолдеров глобальная для всего запроса.
                let mut sql = format!("INSERT INTO {} ({}) VALUES ", qtable, cols);
                let mut param_idx = 1;
                for (r, _) in batch.iter().enumerate() {
                    if r > 0 { sql.push_str(", "); }
                    sql.push('(');
                    for c in 0..ncols {
                        if c > 0 { sql.push_str(", "); }
                        write!(sql, "${}", param_idx).unwrap();
                        param_idx += 1;
                    }
                    sql.push(')');
                }

                let params: Vec<&(dyn postgres::types::ToSql + Sync)> = batch.iter()
                    .flat_map(|row| row.iter()
                        .map(|v| v as &(dyn postgres::types::ToSql + Sync)))
                    .collect();

                let mut tx = self.client.transaction()?;
                tx.execute(sql.as_str(), &params)?;
                tx.commit()?;
                total += batch.len();
                batch.clear();
            }
        }

        Ok(total)
    }

    fn file_insert(&mut self, csv_file: &str, table: &str) -> Result<usize> {
        // Считаем строки за один поток (без загрузки файла в память).
        let row_count = {
            let file = std::fs::File::open(csv_file)
                .map_err(|e| anyhow!("open: {}", e))?;
            let reader = std::io::BufReader::new(file);
            reader.lines().count().saturating_sub(1)
        };

        let copy_sql = format!(
            "COPY {} FROM STDIN WITH (FORMAT csv, HEADER true)",
            Self::quote(table)
        );

        let mut writer = self.client
            .copy_in(&copy_sql)
            .map_err(|e| anyhow!("COPY error: {}", e))?;

        let mut file = std::fs::File::open(csv_file)
            .map_err(|e| anyhow!("open file: {}", e))?;
        std::io::copy(&mut file, &mut writer)
            .map_err(|e| anyhow!("COPY write error: {}", e))?;

        writer.finish()
            .map_err(|e| anyhow!("COPY finish error: {}", e))?;

        Ok(row_count)
    }
}
